NativeLabelScanStore works with consistency checker

tinwelint opened the pull request:
Which means reducing LabelScanStore (and friends) interface and to some extent consistency checker to consistency check label scan store slightly differently.

tinwelint committed:
CC uses KernelExtensions to figure out LSS

tinwelint committed:
Removes LabelScanReader#labelsForNode

since not all implementations can support this efficiently and it was only
used in consistency checking where it was actually redundant anyway.

tinwelint committed:
Implements AllEntriesLabelScanReader for Native LSS

uses aggregation of multiple cursors, one for each label to provide
all label information for each node id range.

This is only used in consistency checker.

tinwelint committed:
Ability to have unknown parts of multi-part ProgressListener

where the unknown parts won't affect total count and instead be seen
as periods where no progress is happening.

tinwelint committed:
Implements LabelScanStore#isEmpty()

and uses it for decision about whether or not to use label scan store to
drive index population, instead of AllEntriesLabelScanReader#maxCount().
This is because not all implementations can know maxCount in advance.

tinwelint commented:
Hmm, not all builds were run on this one

chrisvest commented:
@tinwelint That's a new feature of our PR builds; to only run tests deemed relevant by some dependency analysis.

MishaDemianenko commented:
@chrisvest tests and builds are different things 🦄  😋

tinwelint commented:
@chrisvest @MishaDemianenko I know, but to run no tests at all.... doooon't think so. Actually there was a hickup in CI where other PRs didn't run properly either. Triggered them manually, now the `Run Linux Tests` is there, which is wasn't at all before

tinwelint commented on community/lucene-index/src/main/java/org/neo4j/kernel/api/impl/labelscan/LuceneLabelScanStore.java:
```diff
@@ -137,6 +137,12 @@ public void start() throws IOException
     @Override
+    public boolean isEmpty() throws IOException
+    {
+        return allNodeLabelRanges().maxCount() == 0;
```
Hmm wait, the `allNodeLabelRanges()` should be in a try-with-resource, will fix that

tinwelint committed:
Fixes issue with node <--> index check reporting wrong inconsistencies

tinwelint commented:
Failed on unrelated flaky test, it's fine.

MishaDemianenko commented on community/consistency-check/src/main/java/org/neo4j/consistency/ConsistencyCheckService.java:
```diff
@@ -173,53 +182,48 @@ public Result runFullConsistencyCheck( final File storeDir, Config tuningConfigu
-                        consistencyCheckerConfig, operationalMode, logProvider )
-                        .build();
+            life.start();
+            IndexStoreView indexStoreView = new NeoStoreIndexStoreView( LockService.NO_LOCK_SERVICE, neoStores );
```
it looks like we do not need IndexStoreView and operationalMode anymore here?

tinwelint replied:
Actually this is a mistake, the IndexStoreView should be used as the dependency into the kernel extensions. However it's not reasonable that CC should be able build the label scan store from store on startup to check consistency of it later, but that's apparently how the code did. I don't think we have a single test which starts a CC on a store w/o label scan store though.

I can restore the old behaviour, i.e. using those dependencies again.

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/helpers/progress/ProgressMonitorFactory.java:
```diff
@@ -104,17 +104,37 @@ private MultiPartBuilder( ProgressMonitorFactory factory, String process )
+            return progress;
+        }
+
+        public ProgressListener progressForUnknownPart( String part )
```
that progress will be completely invisible on a output correct?

tinwelint replied:
Yes, we'll have to see if we can do better later on, OK

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/index/labelscan/LabelScanValue.java:
```diff
@@ -32,6 +32,16 @@
      */
     long bits;
 
+    LabelScanValue( long bits )
```
it looks like we do not really need to have this constructor it's used only internally in this class

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/index/labelscan/NativeAllEntriesLabelScanReader.java:
```diff
@@ -0,0 +1,258 @@
+            this.nodes = new long[RANGE_SIZE];
+        }
+
+        void setdRange( long idRange, long[][] labels, int slots )
```
`setdRange` have or missing `I` or additional `d`

tinwelint replied:
Gone now, see comment above.

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/index/labelscan/NativeAllEntriesLabelScanReader.java:
```diff
@@ -0,0 +1,258 @@
+            try
+            {
+                // One "rangeSize" range at a time
+                for ( RawCursor<Hit<LabelScanKey,LabelScanValue>,IOException> cursor : cursors )
```
hm...is it really safe to share cursors between iterator and reader? next call to iterator will gonna have influence on the first one?

tinwelint replied:
Don't think cursors are shared. Iterator is responsible for moving cursors (those that are on the current range) forwards and cram their range data into a NodeLabelRange, which is returned out from the iterator.

Although this comment reminded me about a fix we did after this PR branch, which is that the NodeLabelRange instances can be processed by multiple threads once they leave the iterator, so changed ReusableNodeLabelRange to not be reused, instead creating new instances every time.

MishaDemianenko replied:
ok

MishaDemianenko commented on community/kernel/src/test/java/org/neo4j/kernel/impl/index/labelscan/NativeAllEntriesLabelScanReaderTest.java:
```diff
@@ -0,0 +1,322 @@
+        }
+    }
+
+    private static long count( Labels[] data )
```
this method is not used

MishaDemianenko commented on community/primitive-collections/src/main/java/org/neo4j/collection/primitive/PrimitiveIntCollections.java:
```diff
@@ -826,4 +827,32 @@ public static boolean contains( int[] values, int candidate )
+     * @param values sorted array of int values.
+     * @return the provided array if no duplicates were found, otherwise a new shorter array w/o duplicates.
+     */
+    public static int[] dedup( int[] values )
```
surprisingly OperationsFacade have method that is almost duplicate of this one

tinwelint replied:
Oh, then that was duplicated from NodeItemHelper, which this `dedup` method was moved from. Will deduplicate that method as well :)

MishaDemianenko commented on community/primitive-collections/src/main/java/org/neo4j/collection/primitive/PrimitiveIntCollections.java:
```diff
@@ -826,4 +827,32 @@ public static boolean contains( int[] values, int candidate )
+        for ( int i = 0; i < values.length; i++ )
+        {
+            int value = values[i];
+            for ( int j = 0; j < unique; j++ )
```
i think we can make second index start from i to remove duplicates?

tinwelint replied:
It goes the other way around, doesn't it, from 0 to `unique` which starts at 0 and grows.

MishaDemianenko commented on community/primitive-collections/src/main/java/org/neo4j/collection/primitive/PrimitiveIntCollections.java:
```diff
@@ -826,4 +827,32 @@ public static boolean contains( int[] values, int candidate )
+     * @param values sorted array of int values.
+     * @return the provided array if no duplicates were found, otherwise a new shorter array w/o duplicates.
+     */
+    public static int[] dedup( int[] values )
```
also please give it a full name `deduplicate` or `unique` or soemthing else :)

MishaDemianenko replied:
this and next comment are also applicable for PrimitiveLongCollections

MishaDemianenko reviewed (changes requested):
Couple of mostly minor comments, otherwise looks really good!

tinwelint committed:
Makes use of the created IndexStoreView when loading kernel extensions in CC

tinwelint committed:
Removed unused constructs

tinwelint committed:
Fixes an issue with concurrent label scan CC checking

where NodeLabelRange instances returned from AllEntriesLabelScanReader
can potentially be processed by multiple threads.
This means that the returned instance cannot be reused internally.

tinwelint committed:
Removes unused methods

tinwelint committed:
Renamed dedup --> deduplicate

MishaDemianenko commented:
@tinwelint it's ready for second round right?

tinwelint commented:
@MishaDemianenko yup, sorry didn't mention that
